pages/learn/deploy/offline-updates.md

```diff
@@ -0,0 +1,273 @@
+
+```bash
+$ balena device register ${app_slug} --uuid ${uuid}
+```
+
+### Download balenaOS image
+
+Download the latest production version of [balenaOS][balenaOS] for the device type initalized in the `device_type` version.
+
+```bash
+$ os_version=$(balena os versions ${device_type} | grep prod | head -n 1 | awk '{print $1}')
+$ tmpimg=$(mktemp).img
+
+$ balena os download ${device_type} \
+  --output ${tmpimg} \
+  --version ${os_version}
+```
+
+### Configure balenaOS image
+
+Next, configure the downloaded image by injecting a [config.json][config-file] file using the [`balena os configure`][balena-os-configure] command. Most common system settings can be (re)specified via [config.json](config-file). To preserve the identify of the pre-existing registered device, the same `uuid` initalized earlier will be used for generating a config.json file. 
+
+```bash
+$ tmpconfig=$(mktemp)
+$ config=$(mktemp)
+
+$ balena config generate \
+    --device ${uuid} \
+    --version ${os_version} \
+    --network ethernet \
+    --appUpdatePollInterval 10 \
+    --output ${tmpconfig} \
+    && cat ${tmpconfig} | jq .
+
+$ cat < ${tmpconfig} \
+    | jq --arg keys "$(cat ${ssh_key})" '. + {os: {sshKeys: [$keys]}}' > "${config}" \
+    && cat < ${config} | jq .
+
+$ balena os configure ${tmpimg} \
+    --app ${app_slug} \
+    --device-type ${device_type} \
+    --version ${os_version} \
+    --config-network ethernet \
+    --config ${config}
+
+$ rm ${tmpconfig}
+$ rm ${config}
+```
+
+### Create and Preload release
+
+Offline updates revolve around the concept of [balena preload][balena-preload]. Preload can be used to flash the balenaOS image and your application release in a single step so your device starts running your application containers as soon as it boots. Preloading removes the need for your devices to download the initial application images directly from balena’s build servers, making it an ideal base for the offline update process. Read more about [preloading a device image][preloading-device-image].
+
+> Important note: [balena preload](https://github.com/balena-io/balena-cli/blob/master/INSTALL-MAC.md#balena-preload) functionality requires Docker with AUFS support
+
+Since preload involves flashing an application release, if the pre-existing application doesn't have any releases then a release is needed to be made. Otherwise this next step is optional. 
+
+```bash
+$ date +%s > app/.epoch
+
+$ balena deploy ${app_slug} --build --source .
+```
+
+The following steps will flash the latest release onto the balenaOS image downloaded in the previous steps along with pinning the device to said release. 
+
+```bash
+$ commit=$(balena app ${app_slug} | grep COMMIT | awk '{print $2}')
+
+$ balena preload ${tmpimg} \
+    --app ${app_slug} \
+    --commit ${commit} \
+    --pin-device-to-release
```

Comment:
For devices with low bandwidth internet connections, the user has probably already pinned the device to the release it is running, so that the device doesn't try to start downloading any new release they create.
Given that, it might actually make sense to change this to `--no-pin-device-to-release` and do the pinning in the `patch state endpoint` script by adding
```
,\"should_be_running__release\":${release_id}
```
Such users will have to run that script then before they boot their device.
What do you think @ab77 ?

Reply:
@thgreasi I really don't like that wall of bash text to patch state endpoints, so I don't really have an opinion. Ideally, we should go with whatever reduced the number of steps or even more ideally, have the CLI perform these as part of the preload flow (if a flag is specified presumably) cc @pdcastro

Reply:
> [...] even more ideally, have the CLI perform these as part of the preload flow (if a flag is specified presumably) 

Yes, sounds good to me. The wall of bash text could be replaced with a balena-cli PR. To decide on the name of the flag, what is a short description of the _purpose_ of patching the state endpoint? Why do we need / want to patch the state endpoint? Things like `--patch-state` or `--fix-state` sound rather generic...

Reply:
But I am also not saying that offline updates should be delayed until more work is done on preload - up to you.

Reply:
> so that the device doesn't try to start downloading any new release they create [as soon as the `preload` command is executed]

This is a good point. But even moving the pinning to the patch state endpoint step (which is currently described as an optional step for low bandwidth devices) would still require coordination with the physical updating of each device in a strict order, for example:

1. Flash the media with the image created by `preload` (without release pinning).
2. Take the device offline for the update (rebooting would do).
3. While the device is offline / before it boots, patch the state endpoint to pin the release.

Now, picture devices in multiple remote sites, where the operators doing the flashing do not have a laptop with them to patch the state endpoint then and there for that device...

There is a conflict here between "really offline" and _"online with low bandwitdh, could download images but don't want to, it would likely fail anyway while wasting bandwidth and money"._ For the latter, perhaps we needed a new supervisor feature like "low bandwith mode - don't ever try to download images, await for offline updates".

Reply:
@ab77 

If the device is really always offline, then it doesn't make any difference if they `--pin-device-to-release` or pin it in the `Patch State Endpoint` or don't pin at all (but in that case there is also the question of why are they paying for balenaCloud at the first place).

On the other hand, as @pdcastro said,
If the device has partial/slow/expensive connectivity, the `--pin-device-to-release` would issue the device to try to fetch the new release. I realize that having to run a curl before booting the new preloaded image doesn't sound great, but:
* it should be do-able for an operator to do that, given that the device already has limited connectivity
* It unlocks an extra use case, which imo sounds like what customers probably need

It would of course be great to have a way to instruct a device to not try to download any new release as Paulo said.